### DISCUSSION TITLE
Add pagination object for every query and request

### DISCUSSION THREAD

**Discussion by vmesel** [prest/prest#774] [API Server] [answered]
As I was developing a major feature at work, I needed to create a pagination widget containing: which page I'm at, the total quantity of pages, and the length of items on the page.

I really think this should be something we must give people the ability to handle with a custom command.

What do you guys think, @avelino, @arxdsilva, @cuducos ?

**Comment by arxdsilva** [comment]
would be a great addition, until now we've only used it via custom queries

**Comment by arxdsilva** [comment]
@vmesel this is also something that will potentially break users, so probably we should consider it for v2

**Comment by vmesel** [comment]
@arxdsilva we could do it optionally with a flag.

**Reply by avelino** [reply]
I think this feature is important (actually necessary)

I would put it in v2 for breaking the api return pattern - we have a lot to improve in version 2

**Comment by c-nv-s** [comment]
I've just encountered this requirement in my app as well and I'm not sure how to work around the limitation.
Surely many more people have met this issue no?

**Reply by arxdsilva** [reply]
@c-nv-s the current work around is to use custom queries

**Reply by c-nv-s** [reply]
@arxdsilva maybe I'm missing something but even in a custom query how does one reshape the json that is output by prest from just the array of results to being wrapped as an object? e.g.
from this:
```
[{"field1":x, "field2":x, "field3":x},{"field1":x,"field2":x,"field3":x}]
```
to this:
```
{
    "total": 2,
    "current_page":1,
    "number_of_pages":1,
    "data":[{"field1":x, "field2":x, "field3":x},{"field1":x,"field2":x,"field3":x}]
 }
 ```
 ???

**Reply by arxdsilva** [reply]
@c-nv-s you'll need to do multiple queries, with something more or less like this:
```sql
with prequery as (
    SELECT * FROM <mytable>
...
    ORDER BY some.priority DESC, {{defaultOrValue "orderby" "created_at desc"}}
    LIMIT {{defaultOrValue "size" "10"}}
    OFFSET({{defaultOrValue "page" "1"}} - 1) * {{defaultOrValue "size" "10"}}
),
query_count AS (
    SELECT (
        SELECT count(*)
        FROM pre_query
    ) AS sum_count
)

SELECT   (SELECT sum_count FROM receipts_query_count) AS count, *
FROM     receipts_query
ORDER BY somequery.priority DESC, {{defaultOrValue "orderby" "created_at desc"}}
```

**Reply by c-nv-s** [reply]
@arxdsilva just wanted to mention that despite your suggestion being a workaround, I noticed it still only adds the summary fields to each element (row) in the array returned by Prest.
The requirement I also needed was a way shape the returned json from Prest so that instead returning an array ``[...]`` as the outermost container, it was an object ``{..., data: [...]}`` that was returned as the outermost container.
I'm assuming this is not possible then? and also is there a solution for this ticket in the v2 branch of the repo?

**Reply by arxdsilva** [reply, marked as answer]
@c-nv-s that was an example of something we had in one of our cases, but you can compose in any way you want, just need to reshape the sql statement, for example this returns in the shape of your requested json:

```sql
WITH paginated_data AS (
  SELECT field1, field2, field3
  FROM your_table
  ORDER BY some_column
  LIMIT 10 OFFSET 0  -- page 1, 10 items per page
),
total_count AS (
  SELECT COUNT(*) AS total FROM your_table
)
SELECT json_build_object(
  'total', total_count.total,
  'current_page', 1,
  'number_of_pages', CEIL(total_count.total / 10.0)::int,
  'data', json_agg(paginated_data)
)
FROM paginated_data, total_count;

```

**Reply by c-nv-s** [reply]
ah ok thanks I never knew you it had to be done from a postgres function or about ``json_build_object``, I thought it would be from a feature within Prest

**Reply by arxdsilva** [reply]
Ah no we actually use them under the hood also. The custom queries are full pg sql compatible, so anything you run in a normal query you can use in it.

**Reply by c-nv-s** [reply]
ok thanks... just fyi the resulting object always seems to get wrapped in an array by Prest so maybe there is a transparent result feature needed there but I'll live with this for now.

**Reply by arxdsilva** [reply]
Oh I'll have a look at that thanks
